File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/ui/profile/ProfilForm.kt
package com.sukashawarma.customer.ui.profile

/**
 * Validasi form Informasi Akun. Cermin PERSIS dari
 * apps/retail-gateway/src/lib/profil.ts -- gateway tetap penentu, validasi di
 * sini hanya supaya pesan salah muncul tanpa menunggu jaringan.
 */
object ProfilForm {
    const val NAMA_MIN = 2
    const val NAMA_MAKS = 60

    fun rapikanNama(masukan: String): String = masukan.trim().replace(Regex("\\s+"), " ")

    /** Pesan salah untuk nama, atau null bila sah. */
    fun galatNama(masukan: String): String? {
        val rapi = rapikanNama(masukan)
        return when {
            rapi.length < NAMA_MIN -> "Nama minimal $NAMA_MIN huruf."
            rapi.length > NAMA_MAKS -> "Nama maksimal $NAMA_MAKS huruf."
            else -> null
        }
    }

    /** Nomor HP Indonesia -> bentuk kanonik 628xxxx, atau null bila tidak wajar. */
    fun normalisasiWhatsApp(masukan: String): String? {
        val bersih = masukan.replace(Regex("[\\s\\-().]"), "")
        val angka = when {
            bersih.startsWith("+62") -> "62" + bersih.substring(3)
            bersih.startsWith("62") -> bersih
            bersih.startsWith("0") -> "62" + bersih.substring(1)
            else -> return null
        }
        return if (Regex("^628\\d{7,12}$").matches(angka)) angka else null
    }

    /** Pesan salah untuk nomor WA, atau null bila sah. Kosong itu sah (tanpa nomor). */
    fun galatWhatsApp(masukan: String): String? {
        if (masukan.isBlank()) return null
        return if (normalisasiWhatsApp(masukan) == null) {
            "Nomor WhatsApp harus nomor HP Indonesia, mis. 0812 3456 7890."
        } else null
    }

    /**
     * Bentuk yang ditampilkan di kolom isian: 628xxxx -> 08xxxx, sebab begitulah
     * orang Indonesia menulis nomornya sendiri. Nilai lain dibiarkan apa adanya.
     */
    fun untukIsian(tersimpan: String?): String {
        val t = tersimpan?.trim().orEmpty()
        return if (t.startsWith("62")) "0" + t.substring(2) else t
    }
}

File: mobile/customer-ios-app/SukaShawarma/UI/Profile/ProfilForm.swift
import Foundation

/// Padanan `ui/profile/ProfilForm.kt`. Cermin PERSIS dari
/// apps/retail-gateway/src/lib/profil.ts — gateway tetap penentu; validasi di
/// sini hanya supaya pesan salah muncul tanpa menunggu jaringan.
enum ProfilForm {
    static let namaMin = 2
    static let namaMaks = 60

    static func rapikanNama(_ masukan: String) -> String {
        masukan.trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
    }

    /// Pesan salah untuk nama, atau nil bila sah.
    static func galatNama(_ masukan: String) -> String? {
        let n = rapikanNama(masukan).count
        if n < namaMin { return "Nama minimal \(namaMin) huruf." }
        if n > namaMaks { return "Nama maksimal \(namaMaks) huruf." }
        return nil
    }

    /// Nomor HP Indonesia → bentuk kanonik 628xxxx, atau nil bila tidak wajar.
    static func normalisasiWhatsApp(_ masukan: String) -> String? {
        let bersih = masukan.replacingOccurrences(of: #"[\s\-().]"#, with: "", options: .regularExpression)
        let angka: String
        if bersih.hasPrefix("+62") { angka = "62" + bersih.dropFirst(3) }
        else if bersih.hasPrefix("62") { angka = bersih }
        else if bersih.hasPrefix("0") { angka = "62" + bersih.dropFirst() }
        else { return nil }
        return angka.range(of: #"^628\d{7,12}$"#, options: .regularExpression) != nil ? angka : nil
    }

    /// Kosong itu sah (tanpa nomor); selain itu harus nomor HP Indonesia.
    static func galatWhatsApp(_ masukan: String) -> String? {
        if masukan.trimmingCharacters(in: .whitespaces).isEmpty { return nil }
        return normalisasiWhatsApp(masukan) == nil
            ? "Nomor WhatsApp harus nomor HP Indonesia, mis. 0812 3456 7890." : nil
    }

    /// 628xxxx → 08xxxx di kolom isian, sebab begitulah orang menulis nomornya sendiri.
    static func untukIsian(_ tersimpan: String?) -> String {
        let t = (tersimpan ?? "").trimmingCharacters(in: .whitespaces)
        return t.hasPrefix("62") ? "0" + t.dropFirst(2) : t
    }
}

/// Padanan `ui/profile/InformasiAkun.kt`.
enum InformasiAkun {
    /// Huruf pertama dari dua kata pertama nama; "SS" bila kosong.
    static func inisial(_ nama: String?) -> String {
        let huruf = (nama ?? "").split(whereSeparator: \.isWhitespace).prefix(2)
            .compactMap { $0.first.map { String($0).uppercased() } }.joined()
        return huruf.isEmpty ? "SS" : huruf
    }

    /// Teks kosong atau hanya spasi dianggap tidak ada.
    static func teksAtauNil(_ nilai: String?) -> String? {
        guard let t = nilai?.trimmingCharacters(in: .whitespaces), !t.isEmpty else { return nil }
        return t
    }
}
